src/_pytest/capture.py

```diff
@@ -413,30 +415,27 @@ def safe_text_dupfile(f, mode, default_encoding="UTF8"):
 class EncodedFile:
     errors = "strict"  # possibly needed by py3 code (issue555)
 
-    def __init__(self, buffer, encoding):
+    def __init__(self, buffer: BinaryIO, encoding: str) -> None:
         self.buffer = buffer
         self.encoding = encoding
 
-    def write(self, obj):
-        if isinstance(obj, str):
-            obj = obj.encode(self.encoding, "replace")
-        else:
+    def write(self, obj: str) -> int:
+        if not isinstance(obj, str):
             raise TypeError(
                 "write() argument must be str, not {}".format(type(obj).__name__)
             )
-        return self.buffer.write(obj)
+        return self.buffer.write(obj.encode(self.encoding, "replace"))
 
-    def writelines(self, linelist):
-        data = "".join(linelist)
-        self.write(data)
+    def writelines(self, linelist: List[str]) -> None:
```

Comment:
typeshed uses Iterable here: https://github.com/python/typeshed/blob/d5851eca6f448b648e1a7b3f3010cc3dde498535/stdlib/3/typing.pyi#L490

Would also change the argument name to just `lines`.